pkg/registry/core/pod/storage/eviction.go

```diff
@@ -226,10 +228,28 @@ func (r *EvictionREST) Create(ctx context.Context, name string, obj runtime.Obje
 		pdb := &pdbs[0]
 		pdbName = pdb.Name
 
-		// If the pod is not ready, it doesn't count towards healthy and we should not decrement
-		if !podutil.IsPodReady(pod) && pdb.Status.CurrentHealthy >= pdb.Status.DesiredHealthy && pdb.Status.DesiredHealthy > 0 {
-			updateDeletionOptions = true
-			return nil
+		if !utilfeature.DefaultFeatureGate.Enabled(kubefeatures.PDBPodHealthyPolicy) {
+			// If the pod is not ready, it doesn't count towards currentHealthy and desiredHealthy and we should not decrement DisruptionsAllowed,
+			// then application guarded by the PDB is not disrupted at the moment and deleting non ready pod will not disrupt it
+			if !podutil.IsPodReady(pod) && pdb.Status.CurrentHealthy >= pdb.Status.DesiredHealthy && pdb.Status.DesiredHealthy > 0 {
+				// make sure the pod status hasn't changed status before we delete
+				updateDeletionOptions = true
+				return nil
+			}
+		} else {
+			// If the user has specified a pod healthy policy, check if the pod is healthy according to the provided policy.
+			podHealthy := podHealthyFunc(pdb)
+			// In case podHealthy function is IsPodReady:
+			//   - If the pod is not ready, it doesn't count towards currentHealthy and desiredHealthy and we should not decrement DisruptionsAllowed,
+			//     then application guarded by the PDB is not disrupted at the moment and deleting non healthy pod will not disrupt it.
+			// In case podHealthy function is IsPodRunning:
+			//   - Then the function will always return true, (otherwise it would have been picked up by canIgnorePDB)
+			//   - Then the pod does always count towards currentHealthy and desiredHealthy and we should decrement DisruptionsAllowed.
+			if !podHealthy(pod) && pdb.Status.CurrentHealthy >= pdb.Status.DesiredHealthy && pdb.Status.DesiredHealthy > 0 {
+				// Make sure the pod status hasn't changed status before we delete.
+				updateDeletionOptions = true
+				return nil
+			}
```

Comment:
This part needs a proper review/discussion.

In the [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-apps/3017-pod-healthy-policy-for-pdb/README.md#changes-to-the-eviction-api) we are proposing 

```
The eviction API will be updated to use podHealthyPolicy to determine whether a pod which is Running but not Ready can be evicted regardless of the value of disruptionsAllowed. This will only be a behavioral change when users have specified a podHealthyPolicy, and will not require the actual API to change.
```

Also, @mortent has used a bit different logic in the original PR: https://github.com/mortent/kubernetes/blob/705331a7476a6a596f3122d9b6e3782ade2bfb8b/pkg/registry/core/pod/storage/eviction.go#L226-L241 to implement it like this.

I do not think this is correct, and have changed it as seen above with additional comments as an explanation what is happening in the code.

In case we would go with @mortent solution, we would evict pods without any regard for PDBs which I think is probably not what we want. I am not seeing a benefit of that. I think we want to guard the application pods with a PDB even with other PodHealthyPolicies..

Reply:
I have changed the code and comments to better describe the required behaviour.